author	Syam Sidhardhan <s.syam@samsung.com>
commit	8151afd39538207ea907857e44d79081b9e0097f
	Fri, 13 Apr 2012 09:37:11 +0000 (12:37 +0300)

gateway: Fix D-Bus memory leak in audio

author	Syam Sidhardhan <s.syam@samsung.com>
commit	ab8a7ebc91fb0a39c2321a98a8cde4bd26a7ff1d
	Fri, 13 Apr 2012 09:38:16 +0000 (12:38 +0300)

gateway: Fix D-Bus reply memory leak

Re-arrange the code to avoid the D-Bus memory leak. Earlier if rfcomm
is not connected, then there was a reply memory leak.

diff -cr bluez-4.99.pan.orig/audio/gateway.c bluez-4.99.pan/audio/gateway.c
*** bluez-4.99.pan.orig/audio/gateway.c	Fri Oct 12 07:30:04 2012
--- bluez-4.99.pan/audio/gateway.c	Mon Oct 15 04:22:12 2012
***************
*** 182,189 ****
  					DBUS_TYPE_UINT16, &gw->version,
  					DBUS_TYPE_INVALID);
  
! 	if (dbus_connection_send_with_reply(dev->conn, msg, &call, -1) == FALSE)
  		return FALSE;
  
  	dbus_pending_call_set_notify(call, notify, dev, NULL);
  	dbus_pending_call_unref(call);
--- 182,192 ----
  					DBUS_TYPE_UINT16, &gw->version,
  					DBUS_TYPE_INVALID);
  
! 	if (dbus_connection_send_with_reply(dev->conn, msg,
! 							&call, -1) == FALSE) {
! 		dbus_message_unref(msg);
  		return FALSE;
+ 	}
  
  	dbus_pending_call_set_notify(call, notify, dev, NULL);
  	dbus_pending_call_unref(call);
***************
*** 595,607 ****
  	if (!device->conn)
  		return NULL;
  
  	reply = dbus_message_new_method_return(msg);
  	if (!reply)
  		return NULL;
  
- 	if (!gw->rfcomm)
- 		return  btd_error_not_connected(msg);
- 
  	gateway_close(device);
  	ba2str(&device->dst, gw_addr);
  	DBG("Disconnected from %s, %s", gw_addr, device->path);
--- 598,610 ----
  	if (!device->conn)
  		return NULL;
  
+ 	if (!gw->rfcomm)
+ 		return  btd_error_not_connected(msg);
+ 
  	reply = dbus_message_new_method_return(msg);
  	if (!reply)
  		return NULL;
  
  	gateway_close(device);
  	ba2str(&device->dst, gw_addr);
  	DBG("Disconnected from %s, %s", gw_addr, device->path);
